porch/controllers/rootsyncrollouts/pkg/controllers/rootsyncrollout/controller.go

```diff
@@ -0,0 +1,384 @@
+					Name:       rsr.Name,
+					UID:        rsr.UID,
+					Controller: &t,
+				},
+			},
+		},
+		Spec: rsdapi.RootSyncDeploymentSpec{
+			Targets: rsdapi.ClusterTargetSelector{
+				Selector: rsr.Spec.Targets.Selector,
+			},
+			PackageRevision: rsdapi.PackageRevisionRef{
+				Name:      pkgRevision,
+				Namespace: rsr.Namespace,
+			},
+		},
+	}
+}
+
+func (r *RootSyncRolloutReconciler) getPackages(ctx context.Context, rollout *v1alpha1.RootSyncRollout) (map[string][]*porchapi.PackageRevision, error) {
+	var packageRevisionList porchapi.PackageRevisionList
+	var opts []client.ListOption
+	if rollout.Spec.Packages.Namespace != "" {
+		opts = append(opts, client.InNamespace(rollout.Spec.Packages.Namespace))
+	}
+	if rollout.Spec.Packages.Selector != nil {
+		selector, err := metav1.LabelSelectorAsSelector(rollout.Spec.Packages.Selector)
+		if err != nil {
+			return nil, err
+		}
+		opts = append(opts, client.MatchingLabelsSelector{Selector: selector})
+	}
+	if err := r.List(ctx, &packageRevisionList, opts...); err != nil {
+		return nil, err
+	}
+
+	packageMap := make(map[string][]*porchapi.PackageRevision)
+	for i := range packageRevisionList.Items {
+		packageRevision := packageRevisionList.Items[i]
+		packageName := packageRevision.Spec.PackageName
+		// TODO: See if we can handle this with a FieldSelector on packagerevisions.
+		if packageRevision.Spec.Lifecycle != porchapi.PackageRevisionLifecyclePublished {
+			continue
+		}
+		if _, found := packageMap[packageName]; !found {
+			packageMap[packageName] = make([]*porchapi.PackageRevision, 0)
+		}
+		packageMap[packageName] = append(packageMap[packageName], &packageRevision)
+	}
+
+	return packageMap, nil
+}
+
+func (r *RootSyncRolloutReconciler) listRootSyncDeployments(ctx context.Context, rsrName, rsrNamespace string) ([]*rsdapi.RootSyncDeployment, error) {
+	var list rsdapi.RootSyncDeploymentList
+	if err := r.List(ctx, &list, client.MatchingLabels{rootSyncRolloutLabel: rsrName}, client.InNamespace(rsrNamespace)); err != nil {
+		return nil, err
+	}
+	var rsds []*rsdapi.RootSyncDeployment
+	for i := range list.Items {
+		item := &list.Items[i]
+		rsds = append(rsds, item)
+	}
+	return rsds, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *RootSyncRolloutReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := v1alpha1.AddToScheme(mgr.GetScheme()); err != nil {
+		return err
+	}
+
+	var pkgRev porchapi.PackageRevision
+
+	r.Client = mgr.GetClient()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.RootSyncRollout{}).
+		Owns(&rsdapi.RootSyncDeployment{}).
+		Watches(
+			&source.Kind{Type: &pkgRev},
+			handler.EnqueueRequestsFromMapFunc(r.findRolloutsForPackageRevision),
+		).
+		Complete(r)
+}
+
+func (r *RootSyncRolloutReconciler) findRolloutsForPackageRevision(pkgRev client.Object) []reconcile.Request {
+	// There is not support for reverse lookup by label: https://github.com/kubernetes/kubernetes/issues/1348
```

Comment:
I think controller-runtime has some ability to build our own indexes, but I agree this is a pain.

Reply:
Yeah, there is support for building indexes based on fields in controller-runtime, but I couldn't find support for indexes on labels.